Predictive IntelliSense (#262)
PowerShell/PowerShell-RFC: 1-Draft/RFC00XX-PredIctive-IntelliSense.md

```

It is possible to unload or uninstall the PSReadLine module to disable Predictive IntelliSense. This
creates a negative impact to customers as it would remove all features of PSReadLine. For this
reason, Predictive IntelliSense has its own PSReadLine configuration to enable and disable
predictions.

### Enabling and Disabling Predictions

By default, the feature Predictive IntelliSense is disabled. The impact of enabling this feature by
default may cause new and existing users to become confused at the shell prompt. The feature is
enabled and disabled with a PSReadLine command Set-PSReadLineOption that is executed at the shell
prompt or in the user’s Profile.

Predictive IntelliSense in PSReadLine 2.1.0-beta2 currently supports the following arguments for
prediction source;

- None - This option disables Predictive IntelliSense
- History - This option uses only the PSReadLine history for predictions

Starting with PSReadLine 2.2.0-beta1, Predictive IntelliSense will support the following arguments
for prediction source;

- None - This option disables Predictive IntelliSense
- History - This option uses only the PSReadLine history for predictions

> - Plugin – This option uses only registered plugin modules for predictions

Reviewer: Sounds like we removed this one

---
RFC: RFCXXXX
Author: Jason Helmick
Status: Draft
Area: Shell
Comments Due: 10/31/2020
---

# Predictive IntelliSense

> Predictive IntelliSense is an addition to the interactive (Shell) experience to assist in command

Reviewer: ```suggestion
Predictive IntelliSense is an addition to the interactive (shell) experience to assist in command
```

discovery and accelerate full command execution. The prediction suggestion appears as colored text
following the user’s cursor. This enables new and experienced users of PowerShell to discover, edit,
and execute full commands based on matching predictions from the user’s history or additional
provider.

Additional providers enhance historic predictions by providing domain specific commands and task
completions. Predictive IntelliSense includes an extension model to support the registration of
additional providers.

Predictions may be displayed in either **InlineView** or **ListView** depending on the user’s
preference. Below, predictions are displayed with InlineView.

![image](./media/pred-wlc.png)

Below, predictions are displayed with a dropdown in ListView.

![image](./media/pred-listview.png)

> ## Motivation

Reviewer: I'd move this whole header to the top.

> Tab Completion has accelerated the success of both new and experienced PowerShell users. New users

Reviewer: ```suggestion
Tab completion has accelerated the success of both new and experienced PowerShell users. New users
```

get the benefit of discovery; seeing available cmdlets and parameters as options while interactively
typing. Experienced users receive the benefit of acceleration; typing less while using the **<tab>** key
to quickly complete a command.

The increasing amount of technology translates to an increase in cmdlets and full command
complexity. Predictive IntelliSense is an addition to the concept of Tab Completion; helping the
user discover, build and edit full commands based on the user’s history or additional plugins.

## Release plan

- Proposed 7.1 GA

>   - Currently in PowerShell 7.1 PReview 7

Reviewer: ```suggestion
  - Currently in PowerShell 7.1 Preview 7
```

>   - Currently in PowerShell 7.1 PReview 7

Author: Clarified release plan

  - Predictive IntelliSense Whole Line Completion
  - Historical Predictions
  - Extension Framework
- Future PSReadLine previews
  - Predictive IntelliSense Whole Line Completion + ListView
  - Support for additional providers using extension framework
  - Previews begin in October with PSReadLine 2.2.0-beta1

## Goals/Non-Goals

 Goals:

- Provide suggestions from PSReadLine history
- The user will be able to enable and disable Predictive IntelliSense
- The user will be able to change between InlineView and ListView
- The user will have keyboard shortcuts to navigate and edit a prediction
- The user will register additional providers when desired
- The user may customize the color of a prediction to support accessibility needs
- The user will be able to search history for similar predictions

Non-goals:

> - Due to limit use, PowerShell Session history is not part of the historical suggestion

Reviewer: ```suggestion
- Due to the relatively short length of a session compared to your PSReadline history,
  PowerShell session history is not part of the historical suggestion
```

- VSCode implementation. We agree this is important for a consistent experience and are working to

>   bring this feature in a future release

Reviewer: ```suggestion
  bring this feature in a future release, but the user experience for VS Code is outside the scope of this RFC.
```

## Specification

> The proposal is to add Predictive IntelliSense and Extension model to improve the user interactive experience.

Reviewer: ```suggestion
The proposal is to add Predictive IntelliSense and an extension model to improve the user interactive experience.
```

### Key Design Considerations Predictive IntelliSense

> Predictive IntelliSense displays a best match command completion from command history. The

Reviewer 2: What does "best" mean here?  Does it mean most recent or most frequent?

Author: most recent -- will update

prediction may change if a better match is found as the user types additional information. The
predicted command may be accepted, edited, or ignored.

![Prediction - Historical](./media/pre-wlc.png)

### Install and Remove Predictive IntelliSense

Predictive IntelliSense is a part of the PSReadLine module. Updating or installing the current beta
release from PSGallery will provide the Predictive IntelliSense feature.

> 

Reviewer: Can we add info here about our commitment (or non-commitment) to working down to Windows PS 5.1?

Author: Yes - I will add information that the historical predictions in PSReadLine, (not the extension framework or plugin support) will be available to be downloaded from PsGallery for Windows Powershell 5.1

The current release is PSReadLine 2.1.0 Beta 2:

```powershell

> Install-Module PSReadLine -RequiredVersion 2.1.0-beta2 -AllowPrerelease

Reviewer 3: This doesn't work on WPS 5.1. What's the WPS equivalent?

Thanks.

Reviewer 4: The install command or the PSReadLine version?

I believe PSReadLine 2.1 supports WinPS 5.1.

The command may not work if your `PowerShellGet` version is too old. You may need to run `Install-Module PowerShellGet` first to get a version that supports `Install-Module -AllowPrerelease`

Reviewer 3: The install command.

It was several undocumented and painful-to-discover steps. The initial errors complain that the version isn't valid. Find-Module agrees and spills a sea of red ink. It was a horrible user experience.

Your normal admin-scripter will stay away. I would have done so if I really had a choice. Unfortunately, I had a long-term crash bug in PSreadline that I wanted the fix to.

Reviewer 4: What's the result of `gmo PowerShellGet -List` on your machine?

Reviewer 3: Now it's

> ModuleType Version    Name
> ---------- -------    ----
> Script     2.2.5      PowerShellGet
> Script     1.0.0.1    PowerShellGet

Before, it was just version 1.0.0.1.

Reviewer 4: Yeah unfortunately because the 1.0.0.1 version is the one that's in Windows it's the one that many users have, but it's not at all easy to push an upgrade of it through Windows like that (improving the UX at the cost of possibly breaking existing workflows in Windows doesn't fly). All the UX improvements come in later versions, so there's the necessary step of using the version in Windows to bootstrap the new version. At least in Windows PowerShell. PowerShell 7 ships with the newer version.

Reviewer 3: I understand.

But you will always find me pushing the "cause" of the admin-scripter. S/he's the person who has to sit down and bang through this trying to figure out "WTH?".  :-)

Reviewer 2: I think there is a separate discussion outside of the RFC to have PowerShellGetv3 in WinPS in the future.

```

> It is possible to remove, delete, the PSReadLine module to disable Predictive IntelliSense. This

Reviewer: ```suggestion
It is possible to unload or uninstall the PSReadLine module to disable Predictive IntelliSense. This
```

creates a negative impact to customers as it would remove all features of PSReadLine. For this
reason, Predictive IntelliSense has its own PSReadLine configuration to enable and disable

> predictions.

Reviewer: This paragraph should be under the header at line 99

### Enabling and Disabling Predictions

> By default, the feature Predictive IntelliSense is disabled. The impact of enabling this feature by

Reviewer 2: It's unclear how users discover this feature if it's disabled by default.  Most users are probably not on Twitter and may not read our blogs.  I wonder if it makes sense that on first import of a new version of PSReadLine we present some text on new features?

Reviewer: Possibly, I think we should brainstorm there, but I do agree it should probably be off by default.

Author: I'm concerned that having predictive text show up without warning is going to confuse users, so I agree it should be off by default. I do like the idea of maybe a notification of new features for first import. "Add assistance at the command line with predictive intellisense here [url]"

default may cause new and existing users to become confused at the shell prompt. The feature is
enabled and disabled with a PSReadLine command Set-PSReadLineOption that is executed at the shell
prompt or in the user’s Profile.

Predictive IntelliSense currently supports four arguments for a prediction source;

* None - This option disables Predictive IntelliSense
* History - This option uses only the PSReadLine history for predictions
* Plugin – This option uses only registered plugin modules for predictions
* HistoryAndPlugin – This option uses both for predictions

> 

Reviewer 5: I mentioned this in separate comment but will add here as well. It would be better to separate out Prediction from Source so that we can use History completion in Most Recently Used order rather than a predictive algorithm. 

Prediction Method Examples:
* None
* AI Version 1
* Experimental new AI
* Cloud AI using Azure or AWS AI (maybe not but at least possible to add)

I guess source can be singular or additive so an array:
* History
* Menu
* Plugin

Is the plugin able to register itself to be accepted into the configuration list? I may want to try 3-4 different sources but end up with 1-2 as my defaults. Maybe have the 3rd-4th for specialized work on remote environments.

Reviewer 2: We should consider using `Frecency` algorithm rather than most recently used: https://github.com/PowerShell/PSReadLine/issues/1808

Reviewer 6: The `Frecency` algorithm mentioned by @SteveL-MSFT takes into account both frequency and how recent the history is.

To enable Predictive IntelliSense, enter the following command in the shell or in the users Profile:

```powershell
Set-PSReadLineOption -PredictionSource History
```

To disable Predictive IntelliSense, enter the following command in the shell or in the users Profile:

```powershell
Set-PSReadLineOption -PredictionSource None
```

### Changing the Prediction View

Predictions are displayed in one of two view depending on the user preference. The default view is InlineView.

* InlineView – This is the default view and displays the prediction inline with the user’s typing.
  This view is similar to other shells Fish and ZSH.
* ListView – ListView provides a dropdown list of predictions below the line the user is typing.
  Users may quickly scan the list, highlight and select the desired prediction.

> 

Reviewer 5: Should be an option for InlineView alongside a secondary view. Inline and List complement each other. Inline can show the suggested completion with List showing multiple options at once. 

Additional Views to consider are 
* None
* Wide - the view `crtl+space` and `Format-Wide` uses
* Menu - VT Sequence Menu that looks like a real intellisense menu that you get in ISE and VS Code Editors
* TerminalIntegration - Long term, there are feature requests to provide native rendering of completion in Windows Terminal. See https://github.com/microsoft/terminal/issues/3121 and https://github.com/microsoft/terminal/issues/6632

These additional views make me think Inline should be separate On/Off with additional views as a separate configuration with None only affecting what's in its list and not Inline setting.

Author: @dragonwolf83 Thank you!  We are open to considering additional views as we progress. I agree that InlineView and ListView can be complimentary - currently when using ListView, as you highlight selections, they are displayed fully on the same line as your cursor, similar to InlineView. Appreciate your thoughts and feedback as we move forward.

Users may change the view at the command line, or in the user’s PowerShell profile.

```powershell

> Set-PSReadLineOption -PredictionViewStyle InlineView

Reviewer 2: Seems like there should be a default view, but also allow key binding to select whichever view you want at runtime similar to how PSReadLine handles tab completion vs menu completion

> Set-PSReadLineOption -PredictionViewStyle InlineView

Author: Added clarity regarding 2.2.0-beta1 and F2

```

### Change the Prediction Color for Accessibility

> By default, Predictions appear on the same line the user is typing in a different customizable

Reviewer: ```suggestion
By default, predictions appear on the same line the user is typing in a different customizable
```

> color. To support Accessibility needs, the prediction color is settable in the shell or users

Reviewer: ```suggestion
color. To support accessibility needs, the prediction color is settable in the shell or user's
```

> Profile.

Reviewer: ```suggestion
profile.
```

> ![pi-color](./media/pi-color.png)

Reviewer 2: The Chrome browser has an interesting color scheme for predicting the URL in the address bar where the predicted completion is highlighted vs greyed out.  I believe the current `-Colors` may already allow specifying both foreground and background color, but maybe we just need an example in the help showing how to set it to that color scheme as we've heard from multiple users that the default grey text is hard to read.

Author: @SteveL-MSFT I've seen the background highlighting on the AZ predictor - will get color code and add an example for discussion.

![pi-color](./media/pi-color2.png)

```powershell
Set-PSReadLineOption -Colors @{ Prediction = '#8A0303'}

> Set-PSReadLineOption -Colors @{ Prediction = '#2F7004'}

Author: Corrected text and included ListView values and indiction of which version of PSReadLine these values are available.

```

Multiple types of color code values are supported in PSReadLine. For more information see
Set-PSReadLineOption in
[PSReadLine](https://docs.microsoft.com/en-us/powershell/module/PSReadline/Set-PSReadlineOption?view=powershell-7)

Examples of different color code values:

```powershell
 # Use a ConsoleColor enum
Set-PSReadLineOption -Colors @{ Prediction = 'DarkRed'}
Set-PSReadLineOption -Colors @{ Prediction = [ConsoleColor]::DarkRed}

 # 24 bit color escape sequence

> Set-PSReadLineOption -Colors @{ Prediction = "$([char]0x1b)[38;5;100m"}

Reviewer 2: Since prediction requires 7.1 for the plugin model, you can use "`e" to specifying ESC char

Author: Leaving as [char]0x1b) for downlevel support at this time.

 # RGB value
Set-PSReadLineOption -Colors @{ Prediction = "#8181f7"}
```

### Key Bindings for Predictions

Key bindings control cursor movement and additional features within the prediction. To support users

> running Predictive IntelliSense on multiple platforms, key bindings are user-settable in the Shell

Reviewer: ```suggestion
running Predictive IntelliSense on multiple platforms, key bindings are user-settable in the shell
```

> and user’s PowerShell profile.

Reviewer: ```suggestion
and a user’s PowerShell profile.
```

Alt-Backspace | ShellBackwardKillWord  | “
Alt-b	        | ShellBackwardWord	     | “
Alt-f	        | ShellForwardWord	     | “
Alt-B	        | SelectShellBackwardWord| “
Alt-F	        | SelectShellForwardWord | “
“ or ‘	      | Profile Code	         | Insert Smart quotes
( or { or [	  | Profile Code	         | Insert matching braces
Backspace	    | Profile Code    	     | Delete previous character
Alt-w	        | Profile Code	         | Save current line in history – don’t execute
Ctrl-V	      | Profile Code    	     | Paste clipboard as a Here string
Alt-(	        | Profile Code	         | Insert parenthesis around selection or entire line
Alt-‘	        | Profile Code	         | Toggle quotes on argument under cursor
Alt-%	        | Profile Code	         | Replace aliases with resolved commands

### Search prediction history based on cursor location

The prediction history from PSReadLine is searchable when the user begins typing into the shell.
Users may search the entire prediction history for commands matching under the current cursor
position. Users may search prediction history based on 'Verb', 'Noun' and 'Parameter' of a cmdlet

> using the up/down arrow keys.

Reviewer 6: Is this section related to the predictive intellisense feature? It sounds like the history searching capability, which exists for a long time now.

Author: I removed reference to prediction from the text to reduce confusion. User's may not have experience searching history in this manner especially with predictive text being shown, so felt it should be described in the RFC. If this still seems redundant and unnessiary to the existing feature in PSReadLine, I can remove the section.

* Searching history from cmdlet verb.
![Verb](./media/pi-arrow-verb.gif)

* Searching histroy from cmdlet noun.
![noun](./media/pi-arrow-noun.gif)

* Searching history from parameter.
![Param](./media/pi-arrow-param.gif)

### Edit predictions before accepting the prediction

Users may navigate prediction tokens to make changes to parameters, arguments, and additional
pipeline commands. The prediction model will update the displayed prediction from the closest
matches in PSReadLine history or added provider.

In the example below, a user navigates the prediction, pausing to change the -Property argument and
receiving an updated prediction.

![arg change](./media/pi-fb.gif)

## Prediction Extension Model

> Additional providers enhance historic predictions by providing domain specific commands and task

Reviewer 2: Should make clear that these providers are registered with PowerShell engine and not PSReadLine

> completions. Predictive IntelliSense includes an extension model to support the registration of

Reviewer 6: It's getting rather confusing where this "Predictive Intellisense" feature is implemented in. For example, is the extension model provided by PSReadLine? (no, it's provided by PowerShell itself, but requiring the host to consume and render the prediction results). We probably should be more clear about it.

additional providers.

## Goals/Non-Goals

 Goals:

* Provide extension model to support additional providers

> * Support the discovery of providers using Get-Subsystem

Reviewer 5: Please consider different noun. Subsystem is way too generic and doesn't make sense for discovery. `Get-IntellisenseProvider` or `Get-IntellisenseSystem` makes more sense. For easier typing, `Get-CompletionProvider` would work too. 

Also, please prefix this with PS so it is `Get-PSIntellisenseProvider`

Reviewer 2: Prediction plugins are using the general PowerShell `subsystem` model, but that cmdlet should probably be called `Get-PSSubsystem`, perhaps have a `-Type` parameter where `Prediction` is a valid type making it easy to filter?  `-Type` can be added later since we don't support other types currently.

Reviewer 6: > but that cmdlet should probably be called `Get-PSSubsystem`

Agreed that we can revisit the cmdlet name.

>  a -Type parameter where Prediction is a valid type making it easy to filter?

It already does so: `Get-Subsystem -Kind CommandPredictor`

Reviewer 6: The renaming to `Get-PSSubsystem` is done.

> * Render predictions from providers within performance guidelines 20ms

Reviewer 6: This is an implementation detail that is subject to change, so let's not put in this RFC. We will need to have a thorough documentation about the prediction plugin.

* Provide Module owners method of registration

Non-goals:

* Register/Unregister-Subsystem cmdlets are not planned because it's targeting binary subsystem
  implementations, which should deal with registration/unregistration via APIs.
* Server-side provider privacy is the responsibility of the provider module

## Extension Model Specification

> ### Predictive Suggestion Contract to Support Addition Providers

Reviewer 6: ```suggestion
### Predictive Suggestion Contract to Support Additional Providers
```

Below is an abbreviated code listing of the API contract to support suggestions from additional
providers.

```csharp
/// The class represents a predictive suggestion generated by a predictor.
public sealed class PredictiveSuggestion
{
/// Gets the suggestion.
public string SuggestionText { get; }

/// Gets the tooltip of the suggestion.
public string? ToolTip { get; }

/// Initializes a new instance of the <see cref="PredictiveSuggestion"/> class.
/// <param name="suggestion">The predictive suggestion text.</param>
public PredictiveSuggestion(string suggestion)
: this(suggestion, toolTip: null)

> ```

Reviewer 6: This is the code to show how a suggestion is represented, but not the contract APIs from the predictor interface. Again, I think we don't touch the implementation details here. Leave the implementation details to the documentation.

### Discovery of Provider Metadata and Registration

Providers are registered automatically when the user installs a provider module. Removal of
registration occurs when the module is removed. Users will need to discover installed providers
along with additional metadata, such as version information to assist users in upgrading and

> diagnosing problems.

Reviewer 6: > Users will need to discover installed providers
along with additional metadata, such as version information to assist users in upgrading and
diagnosing problems.

I'm not very clear about what this part is about.

To list the currently installed providers and their respective metadata:

```powershell

> Get-SubSystem

Reviewer 2: Show sample output (what the table will look like)

Reviewer 6: You need to use `Get-Subsystem -Kind CommandPredictor` to get the registered prediction plugins.
Note that the `CommandPredictor` is just one subsystem.

Author: Added sample output with example Get-Subsystem -Kind CommandPredictor

```

> ### Expected Provider Performance

Reviewer 6: Shall we discuss the provider perf expectation here? I think it's better discussed in the documentation of the prediction plugin.

Author: Agreed and will remove

Additional providers often will communicate with an internet connect service to provide suggestions.
To minimize the appearance of slow responses, we estimate that the maximum roundtrip time of 20ms is
permissible to not negatively impact users.

If the roundtrip time is exceeded, the predictor model will provide historical suggestions, or
suggestions from another provider that completed in the allowed time.

* Default required provider performance = 20ms

> The current roundtrip time in PSReadLine using the Tab Completion feature is approximately **2-3ms**. We

Reviewer 2: We are making this user configurable?

Reviewer 6: > The current roundtrip time in PSReadLine using the Tab Completion feature is approximately **2-3ms**

Did you measure the tab completion?
My measurement doesn't have tab completion involved, and the 2-3 ms is just typical typing+rendering in PSReadLine. Tab completion usually takes way much longer, ~250 ms or more, especially when there are custom argument completer registered.

Author: AS discussed with @daxian-dbw, this section is removed from the RFC

believe that this current default roundtrip time for additional providers may still exceed user
expectations. With additional testing and feedback, we may need to reduce the roundtrip time to less
than **20ms**.

> ### Registration for Module Owners

Reviewer 2: Need to update to include the data sent back from PSReadLine to the provider to indicate acceptance of a prediction so that the provider can refine it's results next time

Author: Discussed with @daxian-dbw and removing this section from this RFC. Will be included with implementation and design details in Dongbo's RFC.

Predictive IntelliSense is an extensible model offering module authors the opportunity to create their own predictor plugins. Customers using Predictive IntelliSense will benefit from increased coverage of domain specific technologies and other helpful predictors.

Module authors writing new predictors will need to register their predictor with the extension
model. Predictor registration is accomplished with the **OnImport** and **OnRemove** methods
executed on module startup and removal. Additional documentation is available for
[IModuleAssemblyInitializer
(OnImport)](https://docs.microsoft.com/en-us/dotnet/api/system.management.automation.imoduleassemblyinitializer?view=powershellsdk-7.0.0)
and

> [IModuleAssemblyCleanup (OnRemove)](https://docs.microsoft.com/en-us/dotnet/api/system.management.automation.imoduleassemblycleanup?view=powershellsdk-7.0.0).

Reviewer 6: Registration is done by calling the subsystem registration API. `OnImport` and `OnRemove` are the places to call the registration API for a binary module.

* Example of registering a predictor using **OnImport**

```csharp
// The initializer to register the predictor implementation at module loading time.

    public class PredictorInitializer : IModuleAssemblyInitializer
    {
        public void OnImport()
        {
            var predictor = new MyPredictor();
            SubsystemManager.RegisterSubsystem<IPredictor,MyPredictor>(predictor);
        }
    }
```

* Example of unregistering a predictor using **OnRemove**

```csharp
// Clean up the registeration when unloading the module.

    public class PredictorCleanup : IModuleAssemblyCleanup
    {
        public void OnRemove(PSModuleInfo psModuleInfo)
        {
            SubsystemManager.UnregisterSubsystem<IPredictor>(MyPredictor.Identifier);
        }
    }
```

In the above example, the predictor implementation is elsewhere in the module code with the type

> name **MyPredictor**. An object of **MyPredictor** is created in **OnImport** and **OnRemove**. To

Reviewer 6: ```suggestion
name **MyPredictor**. An object of **MyPredictor** is created in **OnImport**. To
```

### Changing the Prediction View

Beginning with 2.2.0-beta1, Predictions are displayed in one of two views depending on the user
preference. The default view is InlineView.

- InlineView – This is the default view and displays the prediction inline with the user’s typing.
  This view is similar to other shells Fish and ZSH.
- ListView – ListView provides a dropdown list of predictions below the line the user is typing.
  Users may quickly scan the list, highlight and select the desired prediction.

> Users may change the view at the command line using the keybinding F2 or `Set-PSReadLineOption. The
> Set-PSReadLineOption may be stored in the user's profile.

Reviewer: ```suggestion
Users may change the view at the command line using the keybinding F2 or `Set-PSReadLineOption`.
The `Set-PSReadLineOption` may be stored in the user's profile.
```

Alt-‘	        | Profile Code	         | Toggle quotes on argument under cursor
Alt-%	        | Profile Code	         | Replace aliases with resolved commands

### Search history based on cursor location

The history from PSReadLine is searchable when the user begins typing into the shell.
Users may search the entire prediction history for commands matching under the current cursor
position. Users may search history based on 'Verb', 'Noun' and 'Parameter' of a cmdlet
using the up/down arrow keys.

- Searching history from cmdlet verb.
![Verb](./media/pi-arrow-verb.gif)

> - Searching histroy from cmdlet noun.

Reviewer: ```suggestion
- Searching history from cmdlet noun.
```
